### PR TITLE
plan: move projection elimination to physical optimizing phase.

### PR DESCRIPTION
**Pull request by hanfei1991**
When we do projection-elimination in logical phase, it will cause some problems during CBO.
e.g. select * from (select id from a) as a join b on a.id = b.id order by a.id;
the "a.id" in order-by's FromID is Projection. When it walk through Projection, it will be replaced by tablescan's id which FromID is TableScan_a. But when we eliminate projection too early, we will miss the replacing phase and cannot recognize this column in tablescan operator. So the should remain the projections before CBO.
@shenli @coocood @zimulala @XuHuaiyu PTAL

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
LGTM

File: plan/eliminate_projection.go [hunk 1]
```diff
@@ -74,7 +68,7 @@ func shallowCopyColumn(colDest, colSrc *expression.Column) *expression.Column {
 // or "SELECT c AS a, c AS b FROM t WHERE d = 1",
 // or "select t1.a, t2.b, t1.b, t2.a from t1, t2 where t1.a < 0 and t2.b > 0".
```
Comment (zimulala) [inline plan/eliminate_projection.go]: Unify the comment.

File: plan/eliminate_projection.go [hunk 2]
```diff
@@ -18,13 +18,13 @@ import (
...
-func EliminateProjection(p LogicalPlan) LogicalPlan {
+func EliminateProjection(p PhysicalPlan) PhysicalPlan {
 	switch plan := p.(type) {
 	case *Projection:
 		if !projectionCanBeEliminated(plan) {
 			break
 		}
-		child := p.GetChildByIndex(0).(LogicalPlan)
+		child := p.GetChildByIndex(0).(PhysicalPlan)
 		// pointer of schema in PROJECTION operator may be referenced by parent operator,
```
Comment (zimulala) [inline plan/eliminate_projection.go]: s/pointer/Pointer

**Comment by zimulala** [comment]
Can you delete the copy on `FromID` and `Position` in `shallowCopyColumn`.

File: plan/eliminate_projection.go [hunk 3]
```diff
@@ -34,20 +34,14 @@ func EliminateProjection(p LogicalPlan) LogicalPlan {
...
 		RemovePlan(p)
 		p = EliminateProjection(child)
-	case *DataSource:
-		// predicates may be pushed down when build physical plan,
-		// so here we copy the schema of Selection to DataSource.
-		if sel, ok := plan.GetParentByIndex(0).(*Selection); ok {
-			plan.SetSchema(sel.GetSchema())
-		}
 	}
 	if len(p.GetChildren()) == 1 {
```
Comment (zimulala) [inline plan/eliminate_projection.go]: Why here need this condition?

**Comment by zimulala** [comment]
LGTM

**Pull request merged**